pkg/devicetwin/dtcontext/dtcontext_test.go

```diff
@@ -0,0 +1,451 @@
+	dtc.DeviceMutex.Store("DeviceID", &testMutex)
+	testMutex.Lock()
+	tests := []struct {
+		name     string
+		deviceID string
+		want     bool
+	}{
+		{
+			name:     "FailureCaseTest",
+			deviceID: "FailureCaseID",
+			want:     false,
+		},
+		{
+			name:     "SuccessCaseTest",
+			deviceID: "DeviceID",
+			want:     true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := dtc.Unlock(tt.deviceID); got != tt.want {
+				t.Errorf("DTContext.Unlock() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+//TestIsDeviceExist is to test IsDeviceExist().
+func TestIsDeviceExist(t *testing.T) {
+	dtc := &DTContext{
+		DeviceList: &sync.Map{},
+	}
+	dtc.DeviceList.Store("SuccessCaseDeviceID", "SuccessCaseValue")
+	tests := []struct {
+		name     string
+		deviceID string
+		want     bool
+	}{
+		{
+			name:     "FailureCase",
+			deviceID: "FailureCaseDeviceID",
+			want:     false,
+		},
+		{
+			name:     "SuccessCase",
+			deviceID: "SuccessCaseDeviceID",
+			want:     true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if got := dtc.IsDeviceExist(test.deviceID); got != test.want {
+				t.Errorf("DTContext.IsDeviceExist() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+//Function TestGetDevice is to test GetDevice().
+func TestGetDevice(t *testing.T) {
+	dtc := &DTContext{
+		DeviceList: &sync.Map{},
+	}
+	var device dttype.Device
+	dtc.DeviceList.Store("FailureCaseDeviceID", "FailureCaseCasevalue")
+	dtc.DeviceList.Store("SuccessCaseDeviceID", &device)
+	tests := []struct {
+		name     string
+		deviceID string
+		want     *dttype.Device
+		wantbool bool
+	}{
+		{
+			//DeviceID not present
+			name:     "FailureCaseNoDeviceID",
+			deviceID: "FailureCaseID",
+			want:     nil,
+			wantbool: false,
+		},
+		{
+			//DeviceID present but unable to get device
+			name:     "FailCaseTestNoDevice",
+			deviceID: "FailureCaseDeviceID",
+			want:     nil,
+			wantbool: false,
+		},
+		{
+			name:     "SuccessCaseTest",
+			deviceID: "SuccessCaseDeviceID",
+			want:     &dttype.Device{},
+			wantbool: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, gotbool := dtc.GetDevice(test.deviceID)
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("DTContext.GetDevice() got = %v, want %v", got, test.want)
+			}
+			if gotbool != test.wantbool {
+				t.Errorf("DTContext.GetDevice() gotbool = %v, want %v", gotbool, test.wantbool)
+			}
+		})
+	}
+}
+
+//Function TestSend is function to test Send().
+func TestSend(t *testing.T) {
+	commChan := make(map[string]chan interface{})
+	testinterface := make(chan interface{}, 1)
+	commChan[dtcommon.TwinModule] = testinterface
+
+	var msg *model.Message
+	dtc := &DTContext{
+		CommChan: commChan,
+	}
+
+	tests := []struct {
+		name     string
+		identity string
+		action   string
+		module   string
+		msg      *model.Message
+		wantErr  bool
+	}{
+		{
+			name:     "FailureCaseTest",
```

Comment:
as commented earlier , all such tests need to have a logical name which describe why this is a failure test. applies to all such table driven tests.